compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -323,12 +324,25 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         }
     }
 
-    /// Check if the two things are equal in the current param_env, using an infcx to get proper
-    /// equality checks.
+    /// Checks whether the predicates of two trait objects are sufficiently equal to allow
+    /// transmutes between them.
+    ///
+    /// This is called on [`ty::PolyExistentialTraitRef`] or [`ty::PolyExistentialProjection`]
+    /// and checks whether there is a subtyping relation between them in either direction.
+    ///
+    /// # Examples
+    ///
+    /// - returns `true` for `dyn for<'a> Trait<fn(&'a u8)>` and `dyn Trait<fn(&'static u8)>`
+    /// - returns `false` for `dyn Trait<for<'a> fn(&'a u8)>` and either of the above
     #[instrument(level = "trace", skip(self), ret)]
-    pub(super) fn eq_in_param_env<T>(&self, a: T, b: T) -> bool
+    pub(super) fn relate_dyn_predicates_bivariantly<T>(
```

Comment:
bivariance is different from being bidirectional subtypes

https://github.com/lcnr/random-rust-snippets/issues/3

change it to `eq_modulo_regions`? :thinking: 

:+1: on manually adding the leak-check here. The alternative would be to instantiate all regions with existentials and do a proper region check at the end, not totally sure whether a leak-check will always be sufficient given that it doesn't have to be catch all errors

Reply:
> bivariance is different from being bidirectional subtypes

Hm, this comment here made me believe that they are supposed to be the same:

https://github.com/rust-lang/rust/blob/336209eef13882bd1e211b24779584cb7ef911eb/compiler/rustc_infer/src/infer/at.rs#L241-L246

I see now that bivariance as implemented does indeed just do nothing, so :+1: on renaming this to avoid confusion.

---

Also, after thinking about this more, I think we can also just structurally compare the predicates (after instantiating the binder + normalizing), no leak check required, because the soundness of `TypeId` requires structural equality anyway (https://github.com/rust-lang/rust/issues/97156).

Is there any reason why this would be a bad idea?

Reply:
> Hm, this comment here made me believe that they are supposed to be the same:

This comment is wrong then :grin: 

> Also, after thinking about this more, I think we can also just structurally compare the predicates (after instantiating the binder + normalizing), no leak check required, because the soundness of TypeId requires structural equality anyway (https://github.com/rust-lang/rust/issues/97156).
>
> Is there any reason why this would be a bad idea?

don't think so, would have some false negatives in generic code, but :shrug:

Reply:
Alright, replaced this with just `==` now.